src/compiler/transformers/es2015.ts

```diff
@@ -4102,87 +4102,22 @@ namespace ts {
-
-            // createAdd will check if each expression binds less closely than binary '+'.
-            // If it does, it wraps the expression in parentheses. Otherwise, something like
-            //    `abc${ 1 << 2 }`
-            // becomes
-            //    "abc" + 1 << 2 + ""
-            // which is really
-            //    ("abc" + 1) << (2 + "")
-            // rather than
-            //    "abc" + (1 << 2) + ""
-            const expression = reduceLeft(expressions, factory.createAdd)!;
-            if (nodeIsSynthesized(expression)) {
-                setTextRange(expression, node);
-            }
-
-            return expression;
-        }
-
-        /**
-         * Gets a value indicating whether we need to include the head of a TemplateExpression.
-         *
-         * @param node A TemplateExpression node.
-         */
-        function shouldAddTemplateHead(node: TemplateExpression) {
-            // If this expression has an empty head literal and the first template span has a non-empty
-            // literal, then emitting the empty head literal is not necessary.
-            //     `${ foo } and ${ bar }`
-            // can be emitted as
-            //     foo + " and " + bar
-            // This is because it is only required that one of the first two operands in the emit
-            // output must be a string literal, so that the other operand and all following operands
-            // are forced into strings.
-            //
-            // If the first template span has an empty literal, then the head must still be emitted.
-            //     `${ foo }${ bar }`
-            // must still be emitted as
-            //     "" + foo + bar
-
-            // There is always atleast one templateSpan in this code path, since
-            // NoSubstitutionTemplateLiterals are directly emitted via emitLiteral()
-            Debug.assert(node.templateSpans.length !== 0);
+            let expression: Expression = factory.createStringLiteral(node.head.text);
+            for (const span of node.templateSpans) {
+                const args = [visitNode(span.expression, visitor, isExpression)];
 
-            const span = node.templateSpans[0];
-            return node.head.text.length !== 0 || span.literal.text.length === 0 || !!length(getLeadingCommentRangesOfNode(span.expression, currentSourceFile));
-        }
+                if (span.literal.text.length > 0) {
+                    args.push(factory.createStringLiteral(span.literal.text));
+                }
 
-        /**
-         * Adds the head of a TemplateExpression to an array of expressions.
-         *
-         * @param expressions An array of expressions.
-         * @param node A TemplateExpression node.
-         */
-        function addTemplateHead(expressions: Expression[], node: TemplateExpression): void {
-            if (!shouldAddTemplateHead(node)) {
-                return;
+                expression = factory.createCallExpression(

```

Comment:
Curious - why create the `concat` inside the loop of template spans instead of outside the loop?

In other words, why does `a${x}b${y}c` emit `"a".concat(x, "b").concat(y, "c")` instead of emitting `"a".concat(x, "b", y, "c")`?

Reply:
@leebyron i believe the observable order of exceptions from ToString calls may require the chained calls; babel had the same change (please double check me)

Reply:
Aha, yes thank you for this.

There's a test included in this change that would break based on that idea


    class C {
        counter: number;
        constructor() {
            this.counter = 0;
        }
        get foo() {
            this.counter++;
            return {
                toString: () => this.counter++,
            };
        }
    }
    const c = new C;
    export const output = \`\${c.foo} \${c.foo}\`;

`"".concat(c.foo).concat(c.foo)` would run: `get foo()`, `toString()`, `get foo()`, `toString()`

and 

`"".concat(c.foo, c.foo)` would run: `get foo()`, `get foo()`, `toString()`, `toString()`

The first one is what you'd expect to see from a template interpolation